### ISSUE TITLE
Question: How best to extend the Document load process 

### ISSUE THREAD

**Issue by DruSatori** [AngleSharp/AngleSharp#475]
In order to add event handlers to the document before the load is completed.

**Comment by DruSatori** [comment]
more context. I'm loading a document and I am trying to ensure that I hand the on load events and such. In my prior implementation, I was handling the load, and processing the JS from the page. I see that those events are not firing yet in AS, so I am looking at where to start working to add that functionality.

**Comment by FlorianRappl** [comment]
Present a MWE to reproduce your problem. The events are firing. I assume you are adding the handler *after* the events have been fired.

In any case see, e.g., [this example](https://github.com/AngleSharp/AngleSharp.Samples/blob/master/AngleSharp.Samples.Demos/Snippets/CustomEventScripting.cs) for event handling.

**Comment by DruSatori** [comment]
I'm actually looking to handle these prior to this, working on some work that you appear to have scheduled. 

handling things like 

`<!DOCTYPE html public "-//W3C//DTD HTML 4.0//en">
<html>
	
	<head>
		<title>Test</title>
		<script type="text/javascript">
		    window.onload = function()
			{
				document.getElementById("TestResult").innerHTML = "Passed";	
			};
			
		</script>
	</head>
	<body>
		<div id="TestName" class="test">
			<p>This test validates that the onload event is 
				triggering as expected.</p>
						
			<p><em>The result is displayed in the below field. The
			default value is 'Failed'</em></p>
		</div>
		
		<div id="TestResult" class="test">
			Failed
		</div>
				
	</body>
</html>`

With the AS setup being the relevant code. OnLoad is never handled today, I don't expect it to be. I'm looking for ideas as to where you would/are planning to implement it. ( so that I can get to work on it now )

`var config = Configuration.Default.WithDefaultLoader(setup => setup.IsResourceLoadingEnabled = true).WithJavaScript().WithCss().WithCookies();
			var address = "http://www.druware.com/unittests/events.onload.html";
			IBrowsingContext context = BrowsingContext.New(config);
			var document = await context.OpenAsync(address);
			var cells = document.Body.QuerySelectorAll("div").Where(m => m.Id == "TestResult" );
			var titles = cells.Select(m => m.TextContent);
			foreach (String title in titles)
			{
				Console.WriteLine("+ " + title.Trim());
			}`

**Comment by DruSatori** [comment]
It looks ( to me, based upon my limited experience with the AS code at least ) that 90% of the foundation is already there. What appears to be missing is the bits of plumbing to have the parser actually trigger the DOMEvents like onload, onunload, etc, and perhaps to provide methods on the document itself to trigger the UX driven events like onclick, etc to call the appropriate handlers from the C# side.

 IT's worth noting, I already have bits of it working using a prexisting model that just uses jint and the anglesharp parser, but it seems silly to have redundant stack replacing the BrowsingContext, History and other bits.

**Comment by FlorianRappl** [comment]
The `load` event is fired in your case.

One think you need to know is that events like this are *asynchronous* (as specified by the W3C). That means that load is fired *after* the document is loaded. This is also after you received the document via `OpenAsync`. Now a new task is waiting to be executed (which will trigger the callback). You are checking against the document too early - the event loop did not spin again. So no changes are made at that point in time.

If you really would have looked at the snippet you should be surprised (there load is fired, but you are not seeing it). Did you run them at all?

**Comment by DruSatori** [comment]
Alright, so I see what I was doing wrong / misunderstanding, but now that I have it, I think I am missing something.

If the page includes any scripts to be run, then  adding the event listener after the await document is done is fine, as the Loaded event fires after the script executes, and the listener gets triggered. However, if there is nothing to cause any delay, the loaded has fired before it gets added to the listener. I may be wrong, but setting a breakpoint in Document.cs on the RaiseDomContentLoaded() demonstrates what I am seeing. 

`			IBrowsingContext context = BrowsingContext.New(config);
			var document = await context.OpenAsync(address);
			document.Loaded += onLoadedData;`

If the address in the above example contains javascript that assigns an unload() handler, the C# assigned handler does get fired (and the corresponding break in Document.cs hits after the document.Loaded handler is assigned).

Does that mean the event listener needs to be added prior to the OpenAsync() by way of a DocumentLoader, or via the BrowsingContext, or is this something I'm just being obtuse about.

**Comment by FlorianRappl** [comment]
No, `OpenAsync` essentially "waits" already until the document is loaded. In C# there is no need to attach an event handler for the loaded event - in fact: it won't get triggered.

Why is that? Because triggering the event is placed *before* `OpenAsync` finished, however, as the callback is handled *asynchronously* the handling may happen *after* `OpenAsync` finished.

TL;DR: In JS everything should work as expected, but in C# you already start after the document has been processed.

Remark: So why is the handler still "exposed" in C# as well? `OpenAsync` is just one way to load a document. One may also start with a fresh (empty) document and do all the loading. This is done *internally* anyway. In the future such ways may be also exposed publicly allowing users to interfere with any point in the stage, e.g., adding event handlers after certain things happened in the DOM. The major issue at the current point in time is that while JS has a clearly defined state (here is the script - blocking the execution of the parser, i.e., the point of the intervention is clearly known) in C# (naturally being placed outside the document) it is more problematic. A current workaround / solution is to do define a C# scripting engine, which then gets access to the DOM at the same point in time as JS would have access.

So instead of doing this

```html
<!-- some parts-->
<script src="foo.js"></script>
<!-- other parts-->
</body>
```

doing that

```html
<!-- some parts-->
<script src="class://MyCallback" type="text/c-sharp"></script>
<!-- other parts-->
</body>
```

two things need to happen for this (both easily doable today):

* Register a loader for `class://` (or any other convention) "URLs"
* Register a scripting service for the `text/c-sharp` mime-type

As an example the callback would only retrieve a fixed text representation of a class implementing a certain interface. The scripting service would use that information to create the class and run the method with the current document.

**Comment by DruSatori** [comment]
Alright, that's exactly what I see and assumed. It is precisely the behavior I see.

Here is why I'm asking. I load page A. It contains nothing that loads async, so the document returned from .OpenAsync() is loaded and completed. Page B however, executes a script in the document.onload event and therefore, as you note, due to the async nature of the DOM, it is not loaded until that script has run, so any additional work to be done from the C# side, has to wait until that unload event fires. 

From a convenience perspective, being able to attach to onloaded for either case would be the easy way to handle either case, but against since there is no way to attach that unloaded event handler on the C# side prior to the openasync call into the factory, that is not viable. The work around would seem to be set the handler then check the document.ReadyState, if that is complete, then remove the handler, and assume there is no unload event forthcoming, remove the handler and begin the work on the result, otherwise, wait for the unload event to fire and then start processing. ( I haven't tested this work around yet, not at my desk ).  In this situation, I would assume there would need to also be a timer  in place to deal with a situation where the event never fires, or takes too long to fire, and the app needs to time out and cancel the request.

**Comment by FlorianRappl** [comment]
I think the ideal way here would be to roll your own implementation of the `IEventLoop` (the default one is quite lightweight and invokes the asynchronous tasks). Then your implementation either

* stays asynchronous, but gives you the ability to be notified once JS code has been run
* runs synchronous to avoid any complicated interleaving

The latter may seem more easy to implement and manage, however, there may be side-effects that I don't know. I would just give it a try.

As far as rolling your own event loop is concerned - in v0.10 it will be somewhat easier. In v0.9.9 you need to remove the existing from the configuration services (its the `Func<IBrowsingContext, IEventLoop>` instance) and add your own.

**Comment by DruSatori** [comment]
Thanks, I'll look into that this evening.  I suspect the next hurdle will be dealing with the random usage of the <body onload="onloadhandler" ...>...</body> syntax. I haven't even started testing that, but I'm not expecting it to work yet. Any thoughts on where to start with that implementation?

**Comment by FlorianRappl** [comment]
The inline scripting is something that is actually handled, but in a non-optimal way. However, optimal or not - if something does not work here then it is more likely the problem of AngleSharp.Scripting.

**Comment by DruSatori** [comment]
excellent, I'll let you know, and as always, anything I work on in this, I'm happy to give back, so where possible I am trying to stick to existing styles and naming conventions.
